File: week3/plot_part2.m
clear;
clc;
close all;




%% 读取

[T32,M32,E32]=read_series(...
"runs/window-l32/series.jsonl");


[T64,M64,E64]=read_series(...
"runs/window-l64/series.jsonl");



%% ======================
% 计算
% ======================




[T32u,M32a,chi32]=analysis(T32,M32,32);


[T64u,M64a,chi64]=analysis(T64,M64,64);

%% M-T

figure

plot(T32u,M32a,'o-')

hold on

plot(T64u,M64a,'s-')

xlabel("Temperature")

ylabel("<M>")

legend("L=32","L=64")

grid on



%% chi-T

figure

plot(T32u,chi32,'o-')

hold on

plot(T64u,chi64,'s-')

xlabel("Temperature")

ylabel("\chi")

legend("L=32","L=64")

grid on



%% Tc

[~,i1]=max(chi32);

[~,i2]=max(chi64);


fprintf("Tc(L32)=%.3f\n",T32u(i1));

fprintf("Tc(L64)=%.3f\n",T64u(i2));

File: week3/scripts/plot_part2.py
import json
import os
import numpy as np
import matplotlib.pyplot as plt


def read_series(folder):

    filename = os.path.join(folder, "series.jsonl")

    data = {}

    with open(filename, "r") as f:
        for line in f:
            row = json.loads(line)

            T = row["T"]
            M = row["M"]

            if T not in data:
                data[T] = []

            data[T].append(M)

    temps = []
    mean_absM = []
    chi = []

    L = None

    for T in sorted(data.keys()):

        M = np.array(data[T])

        # 从第一行恢复L
        if L is None:
            L = row["L"]

        m_abs = np.mean(np.abs(M))

        susceptibility = (
            L**2 *
            (
                np.mean(M**2)
                -
                np.mean(np.abs(M))**2
            )
            /
            T
        )

        temps.append(T)
        mean_absM.append(m_abs)
        chi.append(susceptibility)

    return (
        np.array(temps),
        np.array(mean_absM),
        np.array(chi)
    )


# ======================
# 读取四组数据
# ======================

runs = {
    "L=32": "artifacts/window-l32",
    "L=64": "artifacts/window-l64"
}


results = {}

for name, folder in runs.items():

    print("Reading", folder)

    results[name] = read_series(folder)



# ======================
# 图1 magnetization
# ======================

plt.figure(figsize=(7,5))


for name, value in results.items():

    T, M, chi = value

    plt.plot(
        T,
        M,
        marker="o",
        label=name
    )


plt.xlabel("Temperature T")
plt.ylabel("<|M|>")

plt.legend()

plt.grid(True)

plt.tight_layout()

plt.savefig(
    "evidence/magnetization.png",
    dpi=300
)

plt.close()



# ======================
# 图2 susceptibility
# ======================


plt.figure(figsize=(7,5))


for name, value in results.items():

    T, M, chi = value

    plt.plot(
        T,
        chi,
        marker="o",
        label=name
    )


plt.xlabel("Temperature T")
plt.ylabel("Susceptibility")


plt.legend()

plt.grid(True)

plt.tight_layout()


plt.savefig(
    "evidence/susceptibility.png",
    dpi=300
)


plt.close()


print("Done!")